### ISSUE TITLE
Nested objects with no ids and different selection sets returns no data

### ISSUE THREAD

**Issue by ford-outreach** [apollographql/apollo-client#7315]
**Intended outcome:**
With returnPartialData false, requesting non-matching selection sets from nested, denormalized objects (with no id), should alert me that there is data loss, and therefore complete data can never be achieved.

In this case, our User has a set of `Preferences`, and under preferences there are `CalendarPreferences`. Neither Preferences nor CalendarPreferences has their own id.

Take this very contrived query (I promise there's a realistic query with the same symptoms)
```
query {
  userWithPreferenceOne: user(id: 1) {
    id
    preferences {
      calendarPreferences {
        preferenceOne
      }
   }
  userWithPreferenceTwo: user(id: 1) {
    id
    preferences {
      calendarPreferences {
        preferenceTwo
      }
   }
 }
}
```

Apollo Client should alert me that because there is no custom merge function defined for Preferences or CalendarPreferences, and because there are no ids, it won't merge them, and therefore either userWithPreferenceOne or userWithPreferenceTwo will never have complete data.

At the very least, [warnAboutDataLoss](https://github.com/apollographql/apollo-client/blob/a975320528d314a1b7eba131b97d045d940596d7/src/cache/inmemory/writeToStore.ts#L334) should fire,
however in a case like this where data will never resolve, an error seems appropriate.

**Actual outcome:**
When a query like the one above is made using the `useQuery` hook. The results are
```
error: undefined,
data: undefined,
loading: false
```

There are no errors or warnings logged, so there is no indication that anything failed. However the query is 'finished' and yet there is no data and no error.

With `returnPartialData` on, the incomplete data is returned (turning this on is undesirable, we need all the data).
Turning on `partialRefetch` leads to an infinite loop of trying to load the complete data and failing.

I stepped through the `warnAboutDataLoss` function and the problematic object (preferences, in the above example) early returns on this condition: https://github.com/apollographql/apollo-client/blob/a975320528d314a1b7eba131b97d045d940596d7/src/cache/inmemory/writeToStore.ts#L362-L365
Since it is only checking top level properties, it verifies that 'existing' and 'incoming' both have 'calendarPreferences' fields, but it does not catch that one only has preferenceOne and one only has preferenceTwo.

**How to reproduce the issue:**
I have created a reproduction here: https://github.com/ford-outreach/react-apollo-error-template
Here is the diff (apologies that it's a little noisy, prettier went to town): https://github.com/apollographql/react-apollo-error-template/compare/master...ford-outreach:master

**Versions**
3.0.2 in our app
3.2.5 in the reproduction

**Comment by benjamn** [comment]
@ford-outreach Your analysis seems spot-on! I fully believe `warnAboutDataLoss` could and should have flagged this problem. We will get that fixed shortly, though I'm also happy to continue discussing the best way to _fix_ these kinds of problems, once they've been flagged by the warning system.

**Comment by ford-outreach** [comment]
Thank you for the quick response @benjamn , that's great to hear! We added ids to Preferences and CalendarPreferences and that's fixed the immediate problem, I think a custom merge function would have also done the trick?

The warning message from warnAboutDataLoss seems super useful, I think had that gone off, it would've been pretty easy to get to the root cause

**Comment by benjamn** [comment]
Yep, the goal of `warnAboutDataLoss` is to get people to either define `keyFields` (or add an `id) or write a custom `merge` function, though I think the first strategy is preferable, when feasible.

**Comment by ford-outreach** [comment]
That makes sense
In this case, the objects in question are sort of pseudo objects at the moment that only exist in the GraphQL layer (hence why they had no ids). Someday they'll have real things with ids backing them, in the meantime we came up with something we can use as a unique identifier in the meantime

**Comment by vieira** [comment]
We were also bitten by this (in 3.3.0-rc.2). We did migrate from 2.x where this used to produce the expected behaviour (most of the time hehe).

As there was no warnings and the query asking for nested object without id was on a different component than the one exhibiting the problem (returning data undefined) it took a while to figure out the issue.

```gql
query GetBook {
  book(id: $id) {
    id
    category {
      # id is missing
      name
    }
    publishedAt
    numPages
  }
```

```gql
query GetBook {
  book(id: $id) {
    id
    reviews {
      edges {
        node {
          text
          createdAt
        }
      }
    }
    publishedAt
    numPages
  }
```

The second query will return `data` undefined. While I understand the problem when two queries ask for category without both asking for the id, I don't understand the issue in this case.

What issue could arise from merging these two results?

Thanks!

**Comment by leethree** [comment]
We encountered the same issue. Agree there should be a warning about this.

We workaround this issue by using `keyFields` to generate IDs:
https://www.apollographql.com/docs/react/caching/cache-configuration/#customizing-identifier-generation-by-type

**Comment by migueloller** [comment]
Another solution instead of writing a customer merge function, is using `merge: true` for the given field as documented [here](https://www.apollographql.com/docs/react/caching/cache-field-behavior/#merging-non-normalized-objects)

**Comment by leethree** [comment]
@migueloller I think `merge: true` works only with objects. In that case there should already be a warning telling users what to do. 

For lists we still need to use either a custom merge function or adding ID/`keyFields` to the list items (as noted by https://github.com/apollographql/apollo-client/issues/7315#issuecomment-726225410). And the warning doesn't show up for some reason.

**Comment by migueloller** [comment]
@leethree, ah, thanks for pointing that out!

**Comment by stephenh** [comment]
@benjamn FYI we had a variation of this (actually an infinite loop of queries, not data loss) that, oddly enough _went away_ in 3.4.0-beta.4 (must be due do that cache canonicalization? I don't see any other commits in there since  3.3.6...).

In our case, for 3.3.6, we had two queries, where one was getting only `keyA` from a `valueObject`:

```
anEntity(id:1) {
  id
  valueObject {
    keyA { id ... }
  }
}
```

And separate query on the same page was getting both `keyA` and `keyB`:

```
anEntity(id:1) {
  id
  valueObject {
    keyA { id ... }
    keyB { id ... }
  }
}
```

(Note that our `valueObject`'s keys actually point to `id`-ized entities, which I think is different from the OP's schema.)

In our case, these two GraphQL queries were invalidating each other, and the `QueryInfo.setDiff` breakpoint's `diff`/`oldDiff` was basically showing:

```
valueObject {
  keyA { ... }
  keyB: undefined
}
```

```
valueObject {
  keyA { ... }
  keyB: null
```

I.e. one query was causing `keyB` to be `undefined` (probably the one that didn't call it) and the other query was causing `keyB` to be `null` (which it was validly `null` from our backend response).

This diff wasn't causing our react components to re-render, however it was causing an ~infinite loop in the cache where each query would take turns invalidating / refetching the other. Here's the Initiator stack trace from one of the ~10s/100s of `graphql` requests in the network tab:

![image](https://user-images.githubusercontent.com/6401/104643393-fcf2dd00-5671-11eb-8440-a9a4d963b2a6.png)

So, odd things:

1) Not sure why in the OP case, it caused data loss, but for us caused an infinite loop of queries, and

   (Actually maybe this is just b/c we have multiple queries on the same page touching the same `anEntity(sameId) { valueObject }` but with different projections.)

2) Not sure why 3.4.0-beta.4 "fixed" it ... i.e. we saw no looping of queries ... is this good/bad/expected?

   I.e. is 3.4.0-beta.4 is merging value objects when it's not supposed to? Or maybe we got lucky and the fact that our diff was "just" `keyB: null` vs. `keyB: undefined`, the canonicalization now threats those as "good enough" the same?

I dunno, maybe this is not a super-helpful update, b/c basically we stayed on 3.3.6 and are using `ValueObject: { merge: true }` in our type policies to fix it, but I nonetheless just wanted to post the FYI that 3.4.0 has a change in behavior here, such that we originally thought "oh this was fixed in 3.4.0-beta.4 and we should just upgrade" ... but actually the real fix is to set `merge: true`, and hence the mystery/question if 3.4.0 caused this change/fix on purpose or not.

Thanks!

**Comment by jdalegonzalez** [comment]
Could @apollo/client/core/ObservableQuery.js not just dump the result of this.queryInfo.getDiff to console.warn if warnAboutDataLoss is true and returnPartialData is false?  Returning void 0 with no error is really unhelpful.  Heck, maybe the error output that's sitting in the diff could be stuck on the error object that's being returned?
